File: handy-js/quiz/q14_group_anagrams_medium.js
/**
 * Q14) Group Anagrams
 *
 * AI-BEST:
 * Canonical key (sorted characters) + hash map groups.
 * Time: O(n * k log k), Space: O(n * k)
 *
 * AI-EASY:
 * Compare each word with existing groups using character counts.
 * Time: O(n^2 * k), Space: O(n * k)
 */

function groupAnagramsBest(words) {
  // Step 1: Build groups using a canonical key.
  // Why: all anagrams share the same sorted-character representation.
  const groupsByKey = new Map();

  for (const word of words) {
    const key = word.split("").sort().join("");
    if (!groupsByKey.has(key)) groupsByKey.set(key, []);
    groupsByKey.get(key).push(word);
  }

  // Step 2: Return grouped values.
  // Why: map values already represent final grouped anagrams.
  return [...groupsByKey.values()];
}

function groupAnagramsEasy(words) {
  // Step 1: Keep a list of groups and place each word by manual comparison.
  // Why: straightforward mental model, even though it is slower.
  const groups = [];

  for (const word of words) {
    let placed = false;
    for (const group of groups) {
      if (isAnagramByCount(word, group[0])) {
        group.push(word);
        placed = true;
        break;
      }
    }

    // Step 2: If no matching group found, start a new one.
    if (!placed) groups.push([word]);
  }

  // Step 3: Return all groups.
  return groups;
}

function isAnagramByCount(a, b) {
  if (a.length !== b.length) return false;

  const count = new Map();
  for (const ch of a) count.set(ch, (count.get(ch) ?? 0) + 1);
  for (const ch of b) {
    const next = (count.get(ch) ?? 0) - 1;
    if (next < 0) return false;
    count.set(ch, next);
  }

  for (const value of count.values()) {
    if (value !== 0) return false;
  }
  return true;
}

function runDemo() {
  console.log("Q14: Group Anagrams");
  const cases = [
    ["eat", "tea", "tan", "ate", "nat", "bat"],
    [""],
    ["a"],
    ["abc", "bca", "cab", "foo", "ofo"],
  ];

  for (const words of cases) {
    console.log(`Input: words=${JSON.stringify(words)}`);
    console.log("  BEST:", JSON.stringify(groupAnagramsBest(words)));
    console.log("  EASY:", JSON.stringify(groupAnagramsEasy(words)));
  }
}

runDemo();

File: handy-python/quiz/q14_group_anagrams_medium.py
"""
Q14) Group Anagrams

AI-BEST:
- Sort each word to build a canonical grouping key.
- Time: O(n * k log k), Space: O(n * k)

AI-EASY:
- Compare each word against existing groups using character counts.
- Time: often O(n^2 * k), Space: O(n * k)
"""

from collections import Counter


def group_anagrams_best(strs: list[str]) -> list[list[str]]:
    # Step 1: Turn each word into a canonical key by sorting its letters.
    # Why: Anagrams share the same sorted character order.
    groups: dict[str, list[str]] = {}
    for word in strs:
        key = "".join(sorted(word))
        groups.setdefault(key, []).append(word)

    # Step 2: Collect the grouped lists from the hash map.
    # Why: The map has already aggregated all anagrams together.
    return list(groups.values())


def group_anagrams_easy(strs: list[str]) -> list[list[str]]:
    # Step 1: Compare each word's character counts to existing groups.
    # Why: This avoids building a canonical sort key and keeps the idea very direct.
    groups: list[list[str]] = []
    signatures: list[Counter[str]] = []

    for word in strs:
        current = Counter(word)
        placed = False

        # Step 2: Reuse an existing group when the counts match.
        # Why: Matching character counts mean the words are anagrams.
        for idx, signature in enumerate(signatures):
            if signature == current:
                groups[idx].append(word)
                placed = True
                break

        # Step 3: Start a new group when no match exists.
        # Why: This preserves the simple linear-scan strategy from the quiz notes.
        if not placed:
            groups.append([word])
            signatures.append(current)

    return groups


def run_demo() -> None:
    print("Q14: Group Anagrams")
    samples = [
        ["eat", "tea", "tan", "ate", "nat", "bat"],
        [""],
        ["abc", "bca", "cab", "foo", "ofo"],
    ]
    for strs in samples:
        print(f"Input: strs={strs}")
        print("  BEST:", group_anagrams_best(strs))
        print("  EASY:", group_anagrams_easy(strs))


if __name__ == "__main__":
    run_demo()
